### PR TITLE
Registration

### LINKED ISSUE #3: Registration
- angular 13
- on authorization dialog you may hide password now
- registration dialog opens
- registration form
- remove useless JWTHelper module
- back registration part

### PR DESCRIPTION
**Pull request by TatsianaMaltseva**
Closes #3

### CODE REVIEW

File: admin/src/app/Components/admin-buttons/admin-buttons.component.html [hunk 1]
```diff
@@ -0,0 +1,7 @@
+<button
+    mat-raised-button
+    (click)="openRegistrationDialog()"
+> 
+    Add new admin 
+</button>
+<!--add hotels + add new facilities + change password-->
```
Comment (bagatka) [inline admin/src/app/Components/admin-buttons/admin-buttons.component.html]: You said you don't like describing tasks in Trello or GitHub Issues, but that's exactly what you do here :)

File: admin/src/app/Components/admin-buttons/admin-buttons.component.ts [hunk 2]
```diff
@@ -0,0 +1,23 @@
+import { Component } from '@angular/core';
+import { MatDialog } from '@angular/material/dialog';
+import { RegistrationComponent } from '../registration/registration.component';
+
+@Component({
+  selector: 'app-admin-buttons',
+  templateUrl: './admin-buttons.component.html',
+  styleUrls: ['./admin-buttons.component.css']
```
Comment (bagatka) [inline admin/src/app/Components/admin-buttons/admin-buttons.component.ts]: Do you need it?

File: admin/src/app/Components/admin-buttons/admin-buttons.component.ts [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+})
+export class AdminButtonsComponent {
+  public constructor(
+    private readonly matDialog: MatDialog) { 
+  }
```
Comment (bagatka) [inline admin/src/app/Components/admin-buttons/admin-buttons.component.ts]:
```suggestion
  public constructor(
    private readonly matDialog: MatDialog
  ) { 
  }
```

File: admin/src/app/Components/authorization/authorization.component.ts [hunk 4]
```diff
@@ -10,14 +10,14 @@ import { AuthorizationDialogComponent } from '../authorizationDialog/authorizati
...
-  }
-
   public constructor(
     private readonly authService: AuthService,
     private readonly matDialog: MatDialog) {
   }
+  
+  public get isLoggedIn(): boolean {
+    return this.authService.isLoggedIn();
+  }
```
Comment (bagatka) [inline admin/src/app/Components/authorization/authorization.component.ts]: It is common for getters to be placed before the constructor. Why did you change the order?

File: admin/src/app/Components/authorization/authorization.component.ts [hunk 5]
```diff
@@ -10,14 +10,14 @@ import { AuthorizationDialogComponent } from '../authorizationDialog/authorizati
   styleUrls: ['./authorization.component.css']
 })
 export class AuthorizationComponent {
-  public get isLoggedIn(): boolean {
-    return this.authService.isLoggedIn();
-  }
-
   public constructor(
     private readonly authService: AuthService,
     private readonly matDialog: MatDialog) {
```
Comment (bagatka) [inline admin/src/app/Components/authorization/authorization.component.ts]:
```suggestion
  public constructor(
    private readonly authService: AuthService,
    private readonly matDialog: MatDialog
  ) {
```

File: admin/src/app/Components/authorizationDialog/authorizationDialog.component.html [hunk 6]
```diff
@@ -1,6 +1,6 @@
 <form mat-dialog-content [formGroup]="authForm">
     <p>Enter your email</p>
-    <mat-form-field appearance="fill">
+    <mat-form-field appearance="fill" style="width: 95%;">
```
Comment (bagatka) [inline admin/src/app/Components/authorizationDialog/authorizationDialog.component.html]: Avoid the use of `style` attribute

File: admin/src/app/Components/authorizationDialog/authorizationDialog.component.html [hunk 7]
```diff
@@ -13,20 +13,28 @@
...
         <mat-error *ngIf="password?.invalid">
             Password is required
         </mat-error>
     </mat-form-field>
     <button mat-button (click)="closeAuthDialog()">Back</button>
     <button 
-    [disabled]="authForm.invalid"
+        [disabled]="authForm.invalid"
         mat-button 
         (click)="login(email?.value, password?.value)"
```
Comment (bagatka) [inline admin/src/app/Components/authorizationDialog/authorizationDialog.component.html]: The order of attributes is incorrect

File: admin/src/app/Components/navbar/navbar.component.html [hunk 8]
```diff
@@ -1,4 +1,5 @@
 <nav>
     <app-authorization></app-authorization>
+    <app-admin-buttons *ngIf="isLoggedIn"></app-admin-buttons>
     <!--other buttons-->
```
Comment (bagatka) [inline admin/src/app/Components/navbar/navbar.component.html]: As Andrew said, don't leave obvious comments

File: admin/src/app/Components/navbar/navbar.component.ts [hunk 9]
```diff
@@ -1,9 +1,17 @@
...
+import { AuthService } from 'src/app/auth.service';
 
 @Component({
   selector: 'app-navbar',
   templateUrl: './navbar.component.html',
   styleUrls: ['./navbar.component.css']
 })
 export class NavbarComponent {
+  public constructor(
+    private readonly authService: AuthService) {
```
Comment (bagatka) [inline admin/src/app/Components/navbar/navbar.component.ts]:
```suggestion
  public constructor(
    private readonly authService: AuthService
  ) {
```

File: admin/src/app/Components/navbar/navbar.component.ts [hunk 10]
```diff
@@ -1,9 +1,17 @@
 import { Component } from '@angular/core';
+import { AuthService } from 'src/app/auth.service';
 
 @Component({
   selector: 'app-navbar',
   templateUrl: './navbar.component.html',
   styleUrls: ['./navbar.component.css']
```
Comment (bagatka) [inline admin/src/app/Components/navbar/navbar.component.ts:7]: Do you need it?

File: admin/src/app/Components/navbar/navbar.component.ts [hunk 11]
```diff
@@ -1,9 +1,17 @@
...
 })
 export class NavbarComponent {
+  public constructor(
+    private readonly authService: AuthService) {
+  }
+  
+  public get isLoggedIn(): boolean {
+    return this.authService.isLoggedIn();
+  }
```
Comment (bagatka) [inline admin/src/app/Components/navbar/navbar.component.ts:12]: Getter should be declared in a different place

File: admin/src/app/Components/registration/registration.component.html [hunk 12]
```diff
@@ -0,0 +1,39 @@
+<h1>Add new admin</h1>
+<form mat-dialog-content [formGroup]="registrationForm">
+    <p>Enter <b>new admin</b> email</p>
+    <mat-form-field apperance="fill" style="width: 95%">
```
Comment (bagatka) [inline admin/src/app/Components/registration/registration.component.html]: Avoid the use of `style` attribute

File: admin/src/app/Components/registration/registration.component.html [hunk 13]
```diff
@@ -0,0 +1,39 @@
...
+        >
+        <mat-error *ngIf="password?.invalid">
+            Password is required
+        </mat-error>
+    </mat-form-field>
+    <button mat-button (click)="closeRegistrationDialog()">Back</button>
+    <button
+        [disabled]="registrationForm.invalid"
+        mat-button 
+        (click)="register(email?.value, password?.value)"
```
Comment (bagatka) [inline admin/src/app/Components/registration/registration.component.html]: Order of attributes is incorrect

File: admin/src/app/Components/registration/registration.component.html [hunk 14]
```diff
@@ -0,0 +1,39 @@
...
+            formControlName="email"
+            type="email"
+            placeholder="new_admin@gmail.com"
+        >
+        <mat-error *ngIf="email?.invalid">
+            Please enter a valid email address
+        </mat-error>
+    </mat-form-field>
+    <p>Enter <b>new admin</b> password</p>
+    <mat-form-field appearance="fill" style="width: 95%;">
```
Comment (bagatka) [inline admin/src/app/Components/registration/registration.component.html]: Avoid use of `style` attribute

File: admin/src/app/Components/registration/registration.component.html [hunk 15]
```diff
@@ -0,0 +1,39 @@
...
+    </mat-form-field>
+    <button mat-button (click)="closeRegistrationDialog()">Back</button>
+    <button
+        [disabled]="registrationForm.invalid"
+        mat-button 
+        (click)="register(email?.value, password?.value)"
+    >
+        Register
+    </button>
+    <mat-error *ngIf="unsuccessWarning">
```
Comment (bagatka) [inline admin/src/app/Components/registration/registration.component.html]: Is there any success warning?

File: admin/src/app/Components/registration/registration.component.ts [hunk 16]
```diff
@@ -0,0 +1,65 @@
...
+})
+export class RegistrationComponent{
+  public registrationForm: FormGroup;
+  public unsuccessWarning: boolean = false;
+
+  public constructor(
+    private readonly authService: AuthService,
+    private readonly matDialogRef: MatDialogRef<RegistrationComponent>,
+    private readonly formBuilder: FormBuilder,
+    private readonly snackBar: MatSnackBar) {
```
Comment (bagatka) [inline admin/src/app/Components/registration/registration.component.ts]:
```suggestion
  public constructor(
    private readonly authService: AuthService,
    private readonly matDialogRef: MatDialogRef<RegistrationComponent>,
    private readonly formBuilder: FormBuilder,
    private readonly snackBar: MatSnackBar
  ) {
```

File: admin/src/app/Components/registration/registration.component.ts [hunk 17]
```diff
@@ -0,0 +1,65 @@
...
+    this.authService.register(email, password)
+    .subscribe(
+      () => {
+        this.unsuccessWarning = false;
+        this.openSuccessSnackBar('Successfully created!');
+      },
+      () => {
+        this.unsuccessWarning = true;
+      }
+    );
```
Comment (bagatka) [inline admin/src/app/Components/registration/registration.component.ts]: I think you decided to use different code style approach for chained methods.
But even with the current choice it is not correct.

Current style
```suggestion
    this.authService.register(email, password)
      .subscribe(
        () => {
          this.unsuccessWarning = false;
          this.openSuccessSnackBar('Successfully created!');
        },
        () => {
          this.unsuccessWarning = true;
        }
      );
```

The style you decided to use
```suggestion
    this.authService
      .register(email, password)
      .subscribe(
        () => {
          this.unsuccessWarning = false;
          this.openSuccessSnackBar('Successfully created!');
        },
        () => {
          this.unsuccessWarning = true;
        }
      );
```

As Andrew has said many times, it doesn't matter which one from correct code style approaches you choose, the important thing is to use it consistently throughout your code base.

File: admin/src/app/auth.service.ts [hunk 18]
```diff
@@ -5,19 +5,20 @@ import { tap } from 'rxjs/operators';
...
   private readonly apiUrl: string;
+  private readonly jwtHelper: JwtHelperService;
   
   public constructor(
-    private readonly http: HttpClient,
-    private readonly jwtHelper: JwtHelperService
+    private readonly http: HttpClient
   ) { 
     this.apiUrl = environment.api;
+    this.jwtHelper = new JwtHelperService();
```
Comment (bagatka) [inline admin/src/app/auth.service.ts]: Why?

File: api/Controllers/AuthController.cs [hunk 19]
```diff
@@ -36,11 +37,31 @@ public IActionResult Login([FromBody]Login request)
...
-            return _hotelsDb.Accounts.SingleOrDefault(u => u.Email == email && u.Password == password);
+            var user = new Account
+            {
+                Email = request.Email,
+                Password = request.Password
+            };
+            if (!IsEmailUnique(user.Email))
+            {
+                return BadRequest("Is already registered with this email");
+            }
```
Comment (bagatka) [inline api/Controllers/AuthController.cs]: Do I need to create a user before checking email uniqueness? If the email is not unique, then memory allocation was useless?

File: api/Controllers/AuthController.cs [hunk 20]
```diff
@@ -36,11 +37,31 @@ public IActionResult Login([FromBody]Login request)
...
+        [Authorize]
+        public IActionResult CreateAccount([FromBody] Login request)
         {
-            return _hotelsDb.Accounts.SingleOrDefault(u => u.Email == email && u.Password == password);
+            var user = new Account
+            {
+                Email = request.Email,
+                Password = request.Password
+            };
+            if (!IsEmailUnique(user.Email))
```
Comment (bagatka) [inline api/Controllers/AuthController.cs]: I think we had a discussion regarding `IsEmailUnique` method naming

File: api/Controllers/AuthController.cs [hunk 21]
```diff
@@ -36,11 +37,31 @@ public IActionResult Login([FromBody]Login request)
...
+            {
+                return BadRequest("Is already registered with this email");
+            }
+            _hotelsDb.Add(user);
+            _hotelsDb.SaveChanges();
```
Comment (bagatka) [inline api/Controllers/AuthController.cs]: I would suggest to use `async` method

File: admin/src/app/Core/getToken.ts [hunk 22]
```diff
@@ -1,5 +1,12 @@
...
 export function getToken(): string | null {
   return localStorage.getItem(ACCESS_TOKEN_KEY);
-}
\ No newline at end of file
+}
+
+export const authorizationHeader = {
+  headers: new HttpHeaders()
+  .set('Authorization', `Bearer ${getToken()}`) 
+};
```
Comment (bagatka) [inline admin/src/app/Core/getToken.ts]: I'm not sure about that solution. Let me investigate it and I'll update you soon

**Review by bagatka — CHANGES_REQUESTED** [review]
Great job!!! I see a huge improvement over the first PR! Just some minor remarks

File: admin/src/app/Components/authorization/authorization.component.ts (hunk 4 above)
Comment (TatsianaMaltseva) [reply admin/src/app/Components/authorization/authorization.component.ts]: image

It to reverse getter and constructor order

File: admin/src/app/Components/navbar/navbar.component.ts (hunk 10 above)
Comment (TatsianaMaltseva) [reply admin/src/app/Components/navbar/navbar.component.ts:7]: I show admin buttons only if user isLoggedIn, so yes I need AuthService

File: api/Controllers/AuthController.cs (hunk 19 above)
Comment (TatsianaMaltseva) [reply api/Controllers/AuthController.cs:53]: Is used to check if token is expired

File: api/Controllers/AuthController.cs (hunk 21 above)
Comment (TatsianaMaltseva) [reply api/Controllers/AuthController.cs]: Hope I changed it right))

File: admin/src/app/auth.service.ts (hunk 18 above)
Comment (TatsianaMaltseva) [reply admin/src/app/auth.service.ts]: Is used to check if token is expired

File: api/Controllers/AuthController.cs (hunk 19 above)
Comment (TatsianaMaltseva) [reply api/Controllers/AuthController.cs]: Genial, why only I didn't think about this point myself

File: admin/src/app/Components/navbar/navbar.component.ts (hunk 10 above)
Comment (bagatka) [reply admin/src/app/Components/navbar/navbar.component.ts:7]: The question was about styles

File: admin/src/app/auth.service.ts (hunk 18 above)
Comment (bagatka) [reply admin/src/app/auth.service.ts]: I mean why did you create it by your own? As I can see, that service was a part of DI

File: admin/src/app/Components/registration/registration.component.html (hunk 15 above)
Comment (TatsianaMaltseva) [reply admin/src/app/Components/registration/registration.component.html]: Was redone with server answer displaying

File: api/Controllers/AuthController.cs [hunk 23]
```diff
@@ -36,11 +38,32 @@ public IActionResult Login([FromBody]Login request)
...
+            _hotelsDb.Add(user);
+            await _hotelsDb.SaveChangesAsync();
+            return NoContent();
         }
 
+        private Account GetAccount(string email, string password) =>
+            _hotelsDb.Accounts.SingleOrDefault(u => u.Email == email && u.Password == password);
+
+        private bool CheckIfEmailUnique(string email) =>
+            _hotelsDb.Accounts.SingleOrDefault(u => u.Email == email) == null;
```
Comment (andrei-kruhliakou-itechart) [inline api/Controllers/AuthController.cs]: Could you think of some other LINQ method that will be more appropriate here?

File: api/Controllers/AuthController.cs [hunk 24]
```diff
@@ -36,11 +38,32 @@ public IActionResult Login([FromBody]Login request)
             return Ok(token);
         }
 
-        private Account GetUserAccount(string email, string password)
+        [Route("registration")]
+        [HttpPost]
+        [Authorize]
```
Comment (andrei-kruhliakou-itechart) [inline api/Controllers/AuthController.cs]: So any user can create an admin user?

File: admin/src/app/Components/admin-buttons/admin-buttons.component.ts [hunk 25]
```diff
@@ -0,0 +1,23 @@
...
+import { RegistrationComponent } from '../registration/registration.component';
+
+@Component({
+  selector: 'app-admin-buttons',
+  templateUrl: './admin-buttons.component.html'
+})
+export class AdminButtonsComponent {
+  public constructor(
+    private readonly matDialog: MatDialog
+    ) {
```
Comment (bagatka) [inline admin/src/app/Components/admin-buttons/admin-buttons.component.ts]:
```suggestion
    private readonly matDialog: MatDialog
  ) { 
```

File: admin/src/app/Components/authorization/authorization.component.ts [hunk 26]
```diff
@@ -6,17 +6,17 @@ import { AuthorizationDialogComponent } from '../authorizationDialog/authorizati
...
   public get isLoggedIn(): boolean {
     return this.authService.isLoggedIn();
   }
-
+  
   public constructor(
     private readonly authService: AuthService,
-    private readonly matDialog: MatDialog) {
+    private readonly matDialog: MatDialog
+    ) {
```
Comment (bagatka) [inline admin/src/app/Components/authorization/authorization.component.ts]:
```suggestion
    private readonly matDialog: MatDialog
  ) {
```

File: admin/src/app/Components/authorizationDialog/authorizationDialog.component.html [hunk 27]
```diff
@@ -13,21 +13,29 @@
...
             formControlName="password"
-            type="password" 
+            [type]="hidePassword ? 'password' : 'text'" 
         >
+        <button 
+            mat-icon-button
+            matSuffix
+            (click)="hidePassword = !hidePassword">
+            <mat-icon>
+                {{hidePassword ? 'visibility_off' : 'visibility'}}
```
Comment (bagatka) [inline admin/src/app/Components/authorizationDialog/authorizationDialog.component.html]:
```suggestion
                {{ hidePassword ? 'visibility_off' : 'visibility' }}
```

File: admin/src/app/Components/registration/registration.component.html [hunk 28]
```diff
@@ -0,0 +1,39 @@
...
+    <button mat-button (click)="closeRegistrationDialog()">Back</button>
+    <button
+        mat-button 
+        [disabled]="registrationForm.invalid"
+        (click)="register(email?.value, password?.value)"
+    >
+        Register
+    </button>
+    <mat-error *ngIf="serverErrorResponse">
+        {{serverErrorResponse}}
```
Comment (bagatka) [inline admin/src/app/Components/registration/registration.component.html]:
```suggestion
        {{ serverErrorResponse }}
```

File: admin/src/app/Components/registration/registration.component.ts [hunk 29]
```diff
@@ -0,0 +1,68 @@
...
+  public get password(): AbstractControl | null {
+    return this.registrationForm.get('password');
+  }
+
+  public constructor(
+    private readonly authService: AuthService,
+    private readonly matDialogRef: MatDialogRef<RegistrationComponent>,
+    private readonly formBuilder: FormBuilder,
+    private readonly snackBar: MatSnackBar
+    ) {
```
Comment (bagatka) [inline admin/src/app/Components/registration/registration.component.ts]:
```suggestion
  ) {
```

**Review by bagatka — CHANGES_REQUESTED** [review]

File: api/Controllers/AuthController.cs (hunk 24 above)
Comment (TatsianaMaltseva) [reply api/Controllers/AuthController.cs]: Roles gonna be added in next pull request

**Review by bagatka — APPROVED** [review]

File: api/Controllers/UsersController.cs [hunk 30]
```diff
@@ -0,0 +1,45 @@
...
+                return BadRequest("User is already registered with this email");
+            }
+            var user = new Account
+            {
+                Email = request.Email,
+                Password = request.Password
+            };
+            _hotelsDb.Add(user);
+            await _hotelsDb.SaveChangesAsync();
+            return NoContent();
```
Comment (andrei-kruhliakou-itechart) [inline api/Controllers/UsersController.cs]: Why 204 status code here?

File: api/Controllers/UsersController.cs [hunk 31]
```diff
@@ -0,0 +1,45 @@
...
+using System.Threading.Tasks;
+
+namespace iTechArt.Hotels.Api.Controllers
+{
+    [Route("api/[controller]")]
+    [ApiController]
+    public class UsersController : Controller
+    {
+        private readonly HotelsDatabaseContext _hotelsDb;
+        private readonly IOptions<UserOptions> _userOptions;
```
Comment (andrei-kruhliakou-itechart) [inline api/Controllers/UsersController.cs]: Why do you need this?

File: admin/src/app/auth.service.ts [hunk 32]
```diff
@@ -43,6 +43,13 @@ export class AuthService {
       );
   }
 
+  public register(email: string, password: string): Observable<string> {
+    return this.http.post<string>(
+      `${this.apiUrl}api/users`,
+      { email, password }
+    );
+  }
+
```
Comment (andrei-kruhliakou-itechart) [inline admin/src/app/auth.service.ts]: Is it actually registration? Why is it in the auth service?

File: api/Controllers/AccountsController.cs [hunk 33]
```diff
@@ -36,10 +34,20 @@ public async Task<IActionResult> CreateAccount([FromBody] Login request)
...
             await _hotelsDb.SaveChangesAsync();
-            return NoContent();
+            return CreatedAtRoute(
+                routeName: "GetLink",
+                routeValues: new { id = user.Id },
+                value: user);
         }
 
+        [Route("{id}")]
+        [HttpGet("{id}", Name = "GetLink")]
```
Comment (andrei-kruhliakou-itechart) [inline api/Controllers/AccountsController.cs]: Why "GetLink"? Why do you even need separate name?

File: api/Controllers/AccountsController.cs [hunk 34]
```diff
@@ -36,10 +34,20 @@ public async Task<IActionResult> CreateAccount([FromBody] Login request)
             };
             _hotelsDb.Add(user);
             await _hotelsDb.SaveChangesAsync();
-            return NoContent();
+            return CreatedAtRoute(
+                routeName: "GetLink",
+                routeValues: new { id = user.Id },
+                value: user);
```
Comment (andrei-kruhliakou-itechart) [inline api/Controllers/AccountsController.cs]: Is it a good practice to use database entities as a return value in controller actions?

File: api/Controllers/AccountsController.cs [hunk 35]
```diff
@@ -36,10 +34,20 @@ public async Task<IActionResult> CreateAccount([FromBody] Login request)
...
         }
 
+        [Route("{id}")]
+        [HttpGet("{id}", Name = "GetLink")]
+        [Authorize]
+        public async Task<Account> GetAccount([FromRoute] int Id) =>
+            await _hotelsDb.Accounts.SingleOrDefaultAsync(account => account.Id == Id);
```
Comment (andrei-kruhliakou-itechart) [inline api/Controllers/AccountsController.cs]: Is it a good practice to use database entities as a return value in controller actions?

File: admin/src/app/Components/registration/registration.component.ts [hunk 36]
```diff
@@ -52,9 +52,9 @@ export class RegistrationComponent{
     );
   }
 
-  public register(email: string, password: string ): void {
-    this.authService
-      .register(email, password)
+  public createAdmin(email: string, password: string ): void {
+    this.userService
+      .createUser(email, password)
```
Comment (andrei-kruhliakou-itechart) [inline admin/src/app/Components/registration/registration.component.ts]: Why 4 different names for one operation with one resource? 
registrationForm - createAdmin - userService - accounts

File: admin/src/app/Components/create-admin/create-admin.component.ts [hunk 37]
```diff
@@ -0,0 +1,68 @@
...
+import { HttpErrorResponse } from '@angular/common/http';
+
+import { AccountService } from 'src/app/account.service';
+
+@Component({
+  selector: 'app-create-admin',
+  templateUrl: './create-admin.component.html',
+  styleUrls: ['./create-admin.component.css']
+})
+export class CreateAdminComponent{
```
Comment (bagatka) [inline admin/src/app/Components/create-admin/create-admin.component.ts]:
```suggestion
export class CreateAdminComponent {
```

File: admin/src/app/Components/create-admin/create-admin.component.ts [hunk 38]
```diff
@@ -0,0 +1,68 @@
...
+  public get password(): AbstractControl | null {
+    return this.registrationForm.get('password');
+  }
+
+  public constructor(
+    private readonly matDialogRef: MatDialogRef<CreateAdminComponent>,
+    private readonly formBuilder: FormBuilder,
+    private readonly snackBar: MatSnackBar,
+    private readonly accountService: AccountService
+    ) {
```
Comment (bagatka) [inline admin/src/app/Components/create-admin/create-admin.component.ts]:
```suggestion
  public constructor(
    private readonly matDialogRef: MatDialogRef<CreateAdminComponent>,
    private readonly formBuilder: FormBuilder,
    private readonly snackBar: MatSnackBar,
    private readonly accountService: AccountService
  ) {
```

File: admin/src/app/Components/create-admin/create-admin.component.ts [hunk 39]
```diff
@@ -0,0 +1,68 @@
...
+        ],
+        password: ['', [Validators.required]]
+      });
+  }
+
+  public closeRegistrationDialog(): void {
+    this.matDialogRef.close();
+  }
+
+  public openSuccessSnackBar(message: string): void {
```
Comment (bagatka) [inline admin/src/app/Components/create-admin/create-admin.component.ts]: Why is it `public`?

File: api/Controllers/AccountsController.cs [hunk 40]
```diff
@@ -0,0 +1,55 @@
...
+            string email = await _hotelsDb.Accounts
+                .Where(account => account.Id == Id)
+                .Select(account => account.Email)
+                .SingleAsync();
+            return Ok(email);
+        }
+
+        private bool CheckIfEmailUnique(string email) =>
+            !_hotelsDb.Accounts.Any(u => u.Email == email);
+
```
Comment (bagatka) [inline api/Controllers/AccountsController.cs]: Remove that line

File: api/Controllers/AccountsController.cs [hunk 41]
```diff
@@ -0,0 +1,55 @@
...
+            };
+            _hotelsDb.Add(user);
+            await _hotelsDb.SaveChangesAsync();
+            return CreatedAtAction(nameof(GetAccountEmailAsync), new { id = user.Id }, null);
+        }
+
+        [Route("{id}")]
+        [HttpGet]
+        [Authorize]
+        public async Task<IActionResult> GetAccountEmailAsync([FromRoute]int Id)
```
Comment (bagatka) [inline api/Controllers/AccountsController.cs]:
```suggestion
        public async Task<IActionResult> GetAccountEmailAsync([FromRoute] int Id)
```

File: api/Controllers/AccountsController.cs [hunk 42]
```diff
@@ -0,0 +1,55 @@
...
+                return BadRequest("User is already registered with this email");
+            }
+            var user = new Account
+            {
+                Email = request.Email,
+                Password = request.Password
+            };
+            _hotelsDb.Add(user);
+            await _hotelsDb.SaveChangesAsync();
+            return CreatedAtAction(nameof(GetAccountEmailAsync), new { id = user.Id }, null);
```
Comment (bagatka) [inline api/Controllers/AccountsController.cs]: Why do you return `null`?

**Review by bagatka — CHANGES_REQUESTED** [review]

File: api/Controllers/AccountsController.cs (hunk 41 above)
Comment (andrei-kruhliakou-itechart) [reply api/Controllers/AccountsController.cs]: Why do you need Async suffix?

File: api/Controllers/AccountsController.cs (hunk 42 above)
Comment (TatsianaMaltseva) [reply api/Controllers/AccountsController.cs]: It is null because I do not need any account data in my angular application now

File: api/Controllers/AccountsController.cs (hunk 41 above)
Comment (TatsianaMaltseva) [reply api/Controllers/AccountsController.cs]: It is naming convention for async methods

Comment (andrei-kruhliakou-itechart) [reply api/Controllers/AccountsController.cs]: What is the purpose of having such convention?

**Review by andrei-kruhliakou-itechart — APPROVED** [review]

**Review by bagatka — APPROVED** [review]
